families/identity/sawtooth_identity/src/handler.rs

```diff
@@ -0,0 +1,460 @@
+      if #[cfg(target_arch = "wasm32")] {
+          use sabre_sdk::ApplyError;
+          use sabre_sdk::TransactionContext;
+          use sabre_sdk::TransactionHandler;
+          use sabre_sdk::TpProcessRequest;
+          use sabre_sdk::{WasmPtr, execute_entrypoint};
+          use identity::{ Policy,
+                          PolicyList,
+                          Role,
+                          RoleList};
+          use setting::Setting;
+
+      } else {
+
+          use sawtooth_sdk::messages::processor::TpProcessRequest;
+          use sawtooth_sdk::processor::handler::ApplyError;
+          use sawtooth_sdk::processor::handler::TransactionContext;
+          use sawtooth_sdk::processor::handler::TransactionHandler;
+          use sawtooth_sdk::messages::setting::Setting;
+          use sawtooth_sdk::messages::identity::{ Policy,
+                                                  PolicyList,
+                                                  Role,
+                                                  RoleList};
+      }
+
+ }
+ use identities::{ IdentityPayload,
+                   IdentityPayload_IdentityType};
+
+ use crypto::digest::Digest;
+ use crypto::sha2::Sha256;
+ use protobuf;
+ use std::collections::HashMap;
+
+  #[cfg(target_arch = "wasm32")]
+  // Sabre apply must return a bool
+  fn apply(
+      request: &TpProcessRequest,
+      context: &mut TransactionContext,
+  ) -> Result<bool, ApplyError> {
+
+      let handler = IdentityTransactionHandler::new();
+      match handler.apply(request, context) {
+          Ok(_) => Ok(true),
+          Err(err) => Err(err)
+      }
+
+  }
+
+  #[cfg(target_arch = "wasm32")]
+  #[no_mangle]
+  pub unsafe fn entrypoint(payload: WasmPtr, signer: WasmPtr) -> i32 {
+      execute_entrypoint(payload, signer, apply)
+  }
+
+
+
+// The identity namespace is special: it is not derived from a hash.
+const IDENTITY_NAMESPACE: &str =  "00001d";
+const ALLOWED_SIGNER_SETTING: &str = "sawtooth.identity.allowed_keys";
+
+// Constants to be used when constructing config namespace addresses
+const SETTING_NAMESPACE: &str = "000000";
+const _SETTING_MAX_KEY_PARTS: usize = 4;
+const _SETTING_ADDRESS_PART_SIZE: usize = 16;
+
+
+/// Computes the address for the given setting key.
+/// Keys are broken into four parts, based on the dots in the string. For
+/// example, the key `a.b.c` address is computed based on `a`, `b`, `c` and
+/// padding. A longer key, for example `a.b.c.d.e`, is still
+/// broken into four parts, but the remaining pieces are in the last part:
+/// `a`, `b`, `c` and `d.e`.
+///
+/// Each of these pieces has a short hash computed (the first
+/// _SETTING_ADDRESS_PART_SIZE characters of its SHA256 hash in hex), and is
+/// joined into a single address, with the config namespace
+/// (_SETTING_NAMESPACE) added at the beginning.
+///
+/// Args:
+///     key (str): the setting key
+/// Returns:
+///     str: the computed address
+///
+
+fn setting_key_to_address(key: &str) -> String {
```

Comment:
This might be a clearer way to implement this: https://github.com/hyperledger/sawtooth-core/blob/master/validator/src/state/settings_view.rs#L128-L145